Addressing issue 89: incorporating SUP on App Registry as OCI Registry (#123)
margo/specification: system-design/specification/applications/application-registry.md

# Application Registry

> The interactions of an `Application Registry` are described [here](../../concepts/applications/application-registry.md) from a conceptual view. This section formally specifies the API of the `Application Registry` and the exchange of an [Application Package](../../concepts/applications/application-package.md), defined through an [Application Description](../../specification/applications/application-description.md) file, from an Application Developer to the [Workload Fleet Manager](../../personas-and-definitions/technical-lexicon.md#workload-fleet-manager) (WFM). The `Application Registry` is designed as an `OCI Registry`, i.e., it offers an API compliant with the [OCI Registry API (v1.1.0)](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md) (called here the "OCI_spec") for digital artifact distribution. This way, the Application Registry hosts the parts of an [Application Package](../../concepts/applications/application-package.md) in the form of `blobs` and uses `image manifests` according to [OCI Image specification](https://github.com/opencontainers/image-spec/blob/v1.0.1/manifest.md) to list a set of `layers`, each pointing at a blob.

Reviewer: The OCI specification has different sub-modules. You are introducing features here available in the OCI Image spec v1.1 and you are referencing that right. For the distribution though the features of the OCI Distribution (API) spec v1.0 suffice. Therefore I would simply leave the version for the OCI Distribution spec "underspecified" (so no version given).

Why? Because although all the registries that I know (GitHub, GitLab, Harbor, ZOT, Docker Registry,...) provide all the needed functionality, only one of them (ZOT) conforms 100% with the OCI Distribution spec v1.1. Someone validating compatibility "on paper" might then think that only one registry can be used, what is not the case.

Author: ok, I removed the version on the "OCI Registry API". Nonetheless, the links to the html document of that version are still there. So that implies a version.

## Overview of API Endpoints

The WFM MUST interact with the Application Registry compliant to the OCI Registry API endpoints defined in the [OCI_spec](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md):

* Tags: `/v2/{name}/tags/list` for listing available versions of an Application Package
* Manifest: `/v2/{name}/manifests/{reference}` for retrieving an image manifest (identifed through the `{reference}`) that lists layers, which represent the parts of the Application Package.
* Blob: `/v2/{name}/blobs/{digest}` for downloading parts of the Application Package

> `{name}` is the namespace of the repository, which needs to be directly communicted by the App Developer to the WFM vendor. It could be for example a combination of the organization's and application's name.

Reviewer: ```suggestion
`{name}` is the namespace of the repository, which needs to be directly communicated by the App Developer to the WFM vendor.
It could be for example a combination of the organization's and application's name.
```

Author: done

Further details on how to use these API endpoints are specified below [towards App Developer](#uploading-an-application-package) and [towards WFM](#retrieving-an-application-package).

## Authentication, Authorization & Security

> Margo recommends the use of an Authentication Service for the interactions between the WFM and Application Registry as conceptually described [here](../../concepts/applications/application-registry.md), e.g., implemented using OAuth 2.0. This involves the following workflow:

Reviewer: ```suggestion
Margo recommends the use of an Authentication Service for the interactions between the WFM and Application Registry as conceptually described [here](../../concepts/applications/application-registry.md), e.g., implemented using OAuth 2.0 (see the [Token Authentication Specification](https://distribution.github.io/distribution/spec/auth/token/) provided by the reference implementation of the spec for more information).
This involves the following workflow:
```

Author: done

* WFM obtains credentials during onboarding
* WFM requests a token from an Authentication Service
* WFM uses the token for subsequent API calls to the Application Registry
* Application Registry validates the token and enforces access control

> Thereby, all communications must use TLS 1.3+ to ensure transport security.

Reviewer: Do we need to fix it here? I'd rather refer to centrally specified minimum security standards in the spec.

Author: I deleted the sentence.

> Note: Authentication, Authorization & Security mechanisms should be defined centrally and homogeneously across Margo. Hence, it is not further defined here.

> Further, tools such as [cosign](https://github.com/sigstore/cosign) may be employed for signing artifacts uploaded to the Application Registry and storing the signatures alongside the artifacts they verify.

Reviewer: ```suggestion
Further, it is recommendable to sign the Application Package.
To do so, the Application Package can be signed as an OCI artifact (with tools such as [cosign](https://github.com/sigstore/cosign)).
Any OCI registry can host those signatures alongside the Application Packages.
Local storing (inside of an archive or in a directory) is also supported over the [OCI Layout](https://github.com/opencontainers/image-spec/blob/main/image-layout.md).
Distributing the signatures alongside the Application Packages they validate can be done with multiple tools available in the OCI ecosystem.
```

Author: done

## Uploading an Application Package

> As shown in the sequence diagram below, the Application Developer uploads the Margo-compliant `Application Package` to the Application Registry. I.e., all parts of the Application Package MUST be pushed as blobs compliant with the [end-4a / end-4b](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints) endpoints of the OCI_spec. 
> 
> Subsequently, the Application Developer creates an OCI image manifest that lists layers, each of which links to an uploaded blob. Then the manifest MUST be pushed to the Application Registry compliant to the [end-7](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints) endpoint of the OCI_spec.
> 
> The uploaded OCI image manifest MUST adhere to the Margo-specific constraints detailed [here](#oci-image-manifest-as-response-from-application-registry).
> 
> Subsequently, the App Developer uses a UI or other vendor-specific mechanism to communicate (either directly or indirectly) to the WFM the namespace of the Application Package's repository. 
> 
> ```mermaid
> sequenceDiagram
> 
>     Note over AppDeveloper: uploads parts of Application Package as blobs:
>     AppDeveloper->>AppRegistry: push blobs
> 
>     AppDeveloper->>AppRegistry: push manifest
> 
>     Note over AppDeveloper: uses vendor-specific upload mechanism (e.g., UI) to enable WFM to find the Application Package:
>     AppDeveloper->>+WFM: Application Package location is: repository name in Application Registry
> ```
> 
> There are no further Margo-specific constraints regarding the upload of the Application Package. The details defined in the [OCI_spec](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md) MUST be applied to this interface of the Application Registry. 
> 
> 
> 

Reviewer: The documentation here is partially incorrect (for example, there are 3 different ways to push a blob and only one of them involves a single call to end-7) and is giving too many technical details that are available in the OCI Distribution spec.

I would restrain from giving too many technical details (e.g. the individual endpoints being used) to reduce  explanation complexity and error probability. I'd rather mention that the detailed workflow is specified there and that multiple tools and libraries support it.

Following is a proposal in that direction for the whole section:

````suggestion
## Uploading an Application Package

This section describes the upload workflow.
However this is an application of the [standard artifact push workflow according the OCI Distribution Specification](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#push) and therefore technical details are to be obtained from that specification.

There are multiple libraries (for example [regclient](https://pkg.go.dev/github.com/regclient/regclient) or [oras](https://pkg.go.dev/oras.land/oras-go/v2) for Go, [oras](https://pypi.org/project/oras/) for Python,...) and tools (for example, [regctl](https://github.com/regclient/regclient), [skopeo](https://github.com/containers/skopeo), [oras](https://github.com/oras-project/oras), [crane](https://github.com/google/go-containerregistry/tree/main/cmd/crane),...) to realize this workflow.

As shown in the sequence diagram below, the Application Developer uploads the Margo-compliant `Application Package` to the Application Registry.
I.e., all parts of the Application Package MUST be pushed as blobs according the ["Pushing Blobs" section of the OCI_spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-blobs).

Subsequently, the Application Developer creates an OCI image manifest that lists layers, each of which links to an uploaded blob.
Then the manifest MUST be pushed to the Application Registry according the ["Pushing Manifests" section of the OCI_spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests).

The uploaded OCI image manifest MUST adhere to the Margo-specific constraints detailed [here](#oci-image-manifest-as-response-from-application-registry).

Subsequently, the App Developer uses a UI or other vendor-specific mechanism to communicate (either directly or indirectly) to the WFM the namespace of the Application Package's repository. 

```mermaid
sequenceDiagram

    Note over AppDeveloper: uploads parts of Application Package as blobs:
    AppDeveloper->>AppRegistry: push blobs

    AppDeveloper->>AppRegistry: push manifest

    Note over AppDeveloper: uses vendor-specific upload mechanism (e.g., UI) to enable WFM to find the Application Package:
    AppDeveloper->>+WFM: Application Package location is: repository name in Application Registry
```
````

Author: done

> ## Retrieving an Application Package
> 
> The WFM has received the namespace of the repository of the Application Package at the Application Registry.
> 
> Next, as shown in the sequence diagram below, the WFM uses the API endpoints defined in the OCI_spec to retrieve a list of Application Package versions (as detailed [here](#list-margo-application-package-versions)).
> 
> Then, the WFM pulls the OCI image manifest of the selected version of the Application Package, with the identifying reference being a tag or digest (as detailed [here](#pull-oci-image-manifest-of-application-package)).
> 
> Finally, all parts (e.g., the Application Description file, the icon, the license, etc.) of the Application Package are retrieved by pulling the respective blobs listed as layers in the image manifest (as detailed [here](#get-parts-of-application-package)). 
> 
> ```mermaid
> sequenceDiagram
>     Note over WFM: retrieve available versions of an Application Package:
>     WFM->>+AppRegistry: GET /v2/{name}/tags/list
>     
>     Note over WFM: retrieves the OCI image manifest of the selected Application Package version. {reference} is a tag or a digest:
>     WFM->>+AppRegistry: GET /v2/{name}/manifests/{reference}
>     
>     AppRegistry-->>+WFM: OCI image manifest
> 
>     Note over WFM: retrieves parts of the Application Package as listed in OCI image manifest layers:
>     WFM->>AppRegistry: GET /v2/{name}/blobs/{digest}
> ```
> 

Reviewer: The same as in the above section. Offloading OCI-specific parts to that specification, keeping this section focused on Margo-only specific parts.

````suggestion
## Retrieving an Application Package

This section describes the retrieval workflow.
However this is an application of the [standard artifact pull workflow according the OCI Distribution Specification](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pull) and therefore technical details are to be obtained from that specification.

The WFM has received the namespace of the repository of the Application Package at the Application Registry.

Next, as shown in the sequence diagram below, the WFM uses the API endpoints defined in the OCI_spec to retrieve a list of Application Package versions (as detailed [here](#list-margo-application-package-versions)).

Then, the WFM pulls the OCI image manifest of the selected version of the Application Package, with the identifying reference being a tag or digest (as detailed [here](#pull-oci-image-manifest-of-application-package)).

If high trust requirements apply on the acquisition of the application package versions, state-of-the-art mechanisms can be applied to ensure that sophisticated attacks are not possible (for example a [TUF](https://theupdateframework.com/) implementation).

Finally, all parts (e.g., the Application Description file, the icon, the license, etc.) of the Application Package are retrieved by pulling the respective blobs listed as layers in the image manifest (as detailed [here](#get-parts-of-application-package)). 

```mermaid
sequenceDiagram
    Note over WFM: retrieve available versions of an Application Package:
    WFM->>+AppRegistry: pull tags list
    
    Note over WFM: retrieves the OCI image manifest of the selected Application Package version. {reference} is a tag or a digest:
    WFM->>+AppRegistry: pull manifest
    
    AppRegistry-->>+WFM: OCI image manifest

    Note over WFM: retrieves parts of the Application Package as listed in OCI image manifest layers:
    WFM->>AppRegistry: pull blobs
```
````

Author: done

The HTTP headers of the ``GET`` request MAY include ```Authorization: Bearer <token>``` for including the interaction with the [Authentication Service](../../concepts/applications/application-registry.md).

To query a subset of tags, the following query parameters MUST be implemented by the Application Registry as defined by OCI_spec endpoint [end-8b](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints):

* ``n=<integer>`` (optional, limits results to n tags)
* ``last=<tagname>`` (optional, starts list of tags after ``<tagname>``)

The response of the HTTP ``GET`` request according to [end-8a/end-8b](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints) is a list of tags, which comply with the versions of available Application Packages. According to OCI_spec, upon success, the response MUST be a JSON body in the following format:

```json
{
  "name": "<name>",
  "tags": [
    "<tag1>", # each tag MUST be the value of the 'metadata.version' of the associated Application Package's Application Description document.
    "<tag2>",
    "<tag3>"
  ]
}
```

> Thereby, a listed tag of an image manifest MUST be the same as the value of the key ``metadata.version`` as specified in the Application Description document of the associated Application Package.

Reviewer: As in the above sections, removing the technical details that should not be provided here and referring to the OCI spec.

I've removed the section mentioning the possibility of having an `Authorization: Bearer <token>` header, because it's an authorization detail that applies to the above mentioned OAuthv2 workflow.

````suggestion
### List Margo Application Package Versions

The interface to retrieve the list of versions of an Application Package MUST be implemented according to OCI_spec endpoint [end-8a](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints). 

The list of the available Application Package versions can be obtained as documented in the ["Listing Tags" section of the OCI_spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-tags).
The API supports filtering and pagination as documented in the linked specification section.

The `{name}` variable is the namespace of the Application Package repository, which was communicated by the App Developer to the WFM vendor.

The list of Application Packages versions is provided in a JSON document that MUST conform with the above mentioned OCI_spec section and therefore has following format:

```json
{
  "name": "<name>",
  "tags": [
    "<tag1>", # each tag MUST be the value of the 'metadata.version' of the associated Application Package's Application Description document.
    "<tag2>",
    "<tag3>"
  ]
}
```

Thereby, a listed tag of an image manifest MUST be the same as the value of the key ``metadata.version`` as specified in the Application Description document of the associated Application Package.
````

Author: done

> 

Reviewer: As in the above sections, removing the technical details that should not be provided here and referring to the OCI spec.

I've removed the section mentioning the possibility of having an `Authorization: Bearer <token>` header, because it's an authorization detail that applies to the above mentioned OAuthv2 workflow.

````suggestion
### List Margo Application Package Versions

The interface to retrieve the list of versions of an Application Package MUST be implemented according to OCI_spec endpoint [end-8a](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints). 

The list of the available Application Package versions can be obtained as documented in the ["Listing Tags" section of the OCI_spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-tags).
The API supports filtering and pagination as documented in the linked specification section.

The `{name}` variable is the namespace of the Application Package repository, which was communicated by the App Developer to the WFM vendor.

The list of Application Packages versions is provided in a JSON document that MUST conform with the above mentioned OCI_spec section and therefore has following format:

```json
{
  "name": "<name>",
  "tags": [
    "<tag1>", # each tag MUST be the value of the 'metadata.version' of the associated Application Package's Application Description document.
    "<tag2>",
    "<tag3>"
  ]
}
```

Thereby, a listed tag of an image manifest MUST be the same as the value of the key ``metadata.version`` as specified in the Application Description document of the associated Application Package.
````

Reviewer: Broken code change suggestion [replaced by a posterior comment](https://github.com/margo/specification/pull/123#pullrequestreview-3539054230).

> #### Example A:

Reviewer: ```suggestion
#### Example A (no filtering):
```

Author: done

Request: GET ``http://famous-app-registry/v2/northstar-industrial-applications/app1/tags/list``

Response: 
```json
{
  "name": "northstar-industrial-applications/app1",
  "tags": [
    "v1.0.0",
    "v1.1.0",
    "latest"
  ]
}
```

> #### Example B:

Reviewer: ```suggestion
#### Example B (filtering to hide versions below v1.0.0):
```

Author: done

Request: GET ``http://famous-app-registry/v2/northstar-industrial-applications/app1/tags/list?n=2&last=v1.0.0``

Response: 
```json
{
  "name": "northstar-industrial-applications/app1",
  "tags": [
    "v1.1.0",
    "latest"
  ]
}
```

### Pull OCI Image Manifest of Application Package

The interface to retrieve the OCI image manifest of a specified version, which belongs to an Application Package, MUST be implemented according to OCI_spec endpoint [end-3](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints).

To pull an Application Package, an HTTP ``GET`` request to a resource path MUST be performed in the following format: 
`/v2/{name}/manifests/{reference}`. 
The `{reference}` MUST be the `tag` or `digest` of an OCI image manifest. The respective `tag` can discovered via the [listing of application package versions](#list-margo-application-package-versions).
The `{name}` variable is the namespace of the Application Package repository.

> The HTTP headers of the ``GET`` request MAY include ```Authorization: Bearer <token>``` for including the interaction with the [Authentication Service](../../concepts/applications/application-registry.md).

Reviewer: Same as above, offloading OCI specifics and authorization details.

```suggestion
### Retrieving OCI Image Manifest of Application Package Manifest

The Application Package OCI Manifest is a JSON document that conforms the [OCI Manifest specification](https://github.com/opencontainers/image-spec/blob/main/manifest.md) and provides metadata required for the distribution of the Application Package (from pushing to pulling) according the OCI specification.

Therefore retrieving an Application Package OCI Manifest MUST be implemented according the ["Pulling manifests" section of the OCISpec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests).

As mentioned before, different libraries and tools do the heavylifting of implementing these workflow.
An implementation without any of those tools or libraries is possible and MUST rely on the OCI_spec to do so.
```

Author: done

        "org.opencontainers.image.title": "resources/license.md"
      }
    },
    {
      "mediaType": "image/jpeg",
      "digest": "sha256:451410b6adfdce1c974da2275290d9e207911a4023fafea0e283fad0502e5e56",
      "size": 5065,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.icon", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/margo.jpg"
      }
    },
    {
      "mediaType": "text/markdown",
      "digest": "sha256:c412d143084c3b051d7ea4b166a7bfffb4550f401d89cae8898991c65e90f736",
      "size": 42,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.releaseNotes", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/release-notes.md"
      }
    }
  ],
}
```

> 

Reviewer: No need to specify the required `schemaVersion` this is not a Margo requirement, but an OCI one.

I'm using "OCI manifest" instead of "OCI image manifest", because although it's specified in the "OCI Image spec", there are no other manifests in the OCI specification.

````suggestion
#### Application Package OCI Manifest

An Application Package OCI Manifest conforms to the [OCI Image Manifest Specification](https://github.com/opencontainers/image-spec/blob/v1.1.0/manifest.md) MUST contain pointers to all parts of an Application Package within the Application Registry.

This section specifies the Margo requirements on these kind of manifests.
Please refer to the [OCI Image Manifest Specification](https://github.com/opencontainers/image-spec/blob/v1.1.0/manifest.md) for all other technical details that are not Margo specific.

Each version of an Application Package MUST have its own OCI manifest.

The ``artifactType`` of the OCI image manifest must be ``application/vnd.margo.app.v1+json``.

The ``config`` MUST be a so-called "empty config" and be specified in the manifest according the ["Guidance for an empty descriptor" section of the OCI Image specification](https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidance-for-an-empty-descriptor).

Each element of the ``layers`` array contains a reference (so-called `digests`) to an artifact (so-called `blobs`) that is a part of the Application Package.

Each Application Package part (file) MUST be listed as an element of the ``layers`` array:

* The [Application Description](../../specification/applications/application-description.md) of the Application Package must be referred to in one element of the ``layers`` array, where the ``mediaType`` of this layer/blob must be ``application/vnd.margo.app.description.v1+yaml``.
* Each [application resource](../../concepts/applications/application-package.md), which is an additional file associated with the application (e.g., manual, icon, release notes, license file, etc.), must be referred to in an element of the ``layers`` array.
  Such a layer must have an ``annotation``, which has the annotation key ``org.margo.app.resource``, and the annotation value must reflect the dotted path to the this resource in the [Application Description file](../../specification/applications/application-description.md).
  E.g., an application icon stored as the file ``resources/margo.jpg`` is referenced in the Application Description under ``metadata.catalog.application.icon: resources/margo.jpg`` and in the OCI image manifest, the respective layer of the icon resource has the annotation ``org.margo.app.resource`` with the value ``metadata.catalog.application.icon`` (see also OCI image manifest example below).
  Future versions of this specification might declare Margo-specific resources (like the application icon, documentation, release-notes,...) and declare [Margo-Specific Media Types](#margo-specific-media-types) for this purpose, replacing the use of annotations for those resources.

The following response example is a Margo-specific OCI manifest following the [OCI Image Manifest Specification](https://github.com/opencontainers/image-spec/blob/v1.0.1/manifest.md) and the above defined specifics:

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.margo.app.v1+json" # this MUST be the artifactType of an OCI image manifest of a Margo Application Package,
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json", # the 'config' object MUST be empty
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30="
  },
  "layers": [
    {
      "mediaType": "application/vnd.margo.app.description.v1+yaml", # this MUST be the artifactType of a Margo Application Description file
      "digest": "sha256:f6b79149e6650b0064c146df7c045d157f3656b5ad1279b5ce9f4446b510bacf",
      "size": 999,
      "annotations": {
        "org.opencontainers.image.title": "margo.yaml"
      }
    },
    {
      "mediaType": "text/markdown",
      "digest": "sha256:e373b123782a2b52483a9124cc9c578e0ed0300cb4131b73b0c79612122b8361",
      "size": 1596,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.descriptionFile", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/description.md"
      }
    },
    {
      "mediaType": "text/markdown",
      "digest": "sha256:af7db4ab9030533b6cda2325247920c3659bc67a7d49f3d5098ae54a64633ec7",
      "size": 25,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.licenseFile", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/license.md"
      }
    },
    {
      "mediaType": "image/jpeg",
      "digest": "sha256:451410b6adfdce1c974da2275290d9e207911a4023fafea0e283fad0502e5e56",
      "size": 5065,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.icon", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/margo.jpg"
      }
    },
    {
      "mediaType": "text/markdown",
      "digest": "sha256:c412d143084c3b051d7ea4b166a7bfffb4550f401d89cae8898991c65e90f736",
      "size": 42,
      "annotations": {
        "org.margo.app.resource": "metadata.catalog.application.releaseNotes", # each Margo application resource MUST be annotated with a dotted path to its definition in the Application Description file
        "org.opencontainers.image.title": "resources/release-notes.md"
      }
    }
  ],
}
```

````

Author: done

> #### Margo-Specific Media Types

Reviewer: As mentioned above, I'd expect future versions of the Margo spec to declare a couple of more Margo-specific resources and declare MediaTypes for them, replacing the use of the annotations for them.

In general filtering based on the value of a JSON document based on a fixed key is much easier and more efficient than filtering for specific keys in a key-value store and then filtering for a specific value.

To better understand it, try to filter with `jq` the above manifest to get the digest of the blob with MediaType `application/vnd.margo.app.description.v1+yaml` and then try again with `jq` to get the digest of the icon.

Author: done

|``application/vnd.margo.app.description.v1+yaml``	| MUST be used to mark a layer in the OCI image manifest as pointing to the Margo Application Description file |

#### Margo-Specific Annotation Keys

|Annotation Key | Description|
|----------|----------|
|``org.margo.app.resource``	| MUST be used to annotate a layer/blob that references a Margo application resource |

### Get Parts of Application Package

The interface to retrieve Application Package parts, i.e., Application Description or Application Resource (e.g., icon, license, etc.), MUST be implemented according to OCI_spec endpoint [end-2](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#endpoints) by pulling a blob.

To pull such a blob, an HTTP ``GET`` request to a resource path MUST be performed in the following format: 
`/v2/{name}/blobs/{digest}`. 
The `{digest}` is the blob's digest as listed in the application's OCI image manifest that has been [retrieved earlier](#pull-oci-image-manifest-of-application-package).
The `{name}` variable is the namespace of the Application Package repository.

> The HTTP headers of the ``GET`` request MAY include ```Authorization: Bearer <token>``` for including the interaction with the [Authentication Service](../../concepts/applications/application-registry.md).

Reviewer: Again as much offloading to the OCI spec as possible and leaving authentication details for a common part.

```suggestion
### Retrieve Parts of Application Package

Retrieving the different files that compose an Application Package MUST be implemented according the ["Pulling blobs" section of the OCI_spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-blobs).

Also for this purpose available tools and libraries can be used for an implementation with low complexity.
```

Author: done
